### PR TITLE
fix #66: add a fix and a test case for ArrayIndexOutOfBoundsException

### LINKED ISSUE #66: ArrayIndexOutOfBoundsException in BufferedSocketInputStream.java:65
Hi @osheroff again

We got another exception in mysql-binlog-connector-java while using Debezium recently😅 This is the full error log we get.

```log
org.apache.kafka.connect.errors.ConnectException: An exception occurred in the change event producer. This connector will be stopped.
    at io.debezium.pipeline.ErrorHandler.setProducerThrowable(ErrorHandler.java:42)
    at io.debezium.connector.mysql.MySqlStreamingChangeEventSource$ReaderThreadLifecycleListener.onEventDeserializationFailure(MySqlStreamingChangeEventSource.java:1193)
    at com.github.shyiko.mysql.binlog.BinaryLogClient.listenForEventPackets(BinaryLogClient.java:958)
    at com.github.shyiko.mysql.binlog.BinaryLogClient.connect(BinaryLogClient.java:606)
    at com.github.shyiko.mysql.binlog.BinaryLogClient$7.run(BinaryLogClient.java:850)
    at java.lang.Thread.run(Thread.java:748)
Caused by: io.debezium.DebeziumException
    at io.debezium.connector.mysql.MySqlStreamingChangeEventSource.wrap(MySqlStreamingChangeEventSource.java:1146)
    ... 5 more
Caused by: java.lang.ArrayIndexOutOfBoundsException
    at com.github.shyiko.mysql.binlog.io.BufferedSocketInputStream.read(BufferedSocketInputStream.java:65)
    at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.readWithinBlockBoundaries(ByteArrayInputStream.java:262)
    at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.read(ByteArrayInputStream.java:241)
    at java.io.InputStream.skip(InputStream.java:224)
    at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.skipToTheEndOfTheBlock(ByteArrayInputStream.java:280)
    at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.deserializeEventData(EventDeserializer.java:305)
    at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.nextEvent(EventDeserializer.java:232)
    at io.debezium.connector.mysql.MySqlStreamingChangeEventSource$1.nextEvent(MySqlStreamingChangeEventSource.java:233)
    at com.github.shyiko.mysql.binlog.BinaryLogClient.listenForEventPackets(BinaryLogClient.java:945)
    ... 3 more
``` 

From the error log, it shows `java.lang.ArrayIndexOutOfBoundsException` in code 

https://github.com/osheroff/mysql-binlog-connector-java/blob/9ec6632c36b567f665b94b407f5eed18f72e2aa5/src/main/java/com/github/shyiko/mysql/binlog/io/BufferedSocketInputStream.java#L65

After digging more about the code, I believe the issue is caused by the value of `limit` in line 

https://github.com/osheroff/mysql-binlog-connector-java/blob/9ec6632c36b567f665b94b407f5eed18f72e2aa5/src/main/java/com/github/shyiko/mysql/binlog/io/BufferedSocketInputStream.java#L62

According to the doc of InputStream, read(...) method will return -1 if it reaches the end of stream, so the `limit` could be -1 here. Then `bytesRemainingInBuffer` will be a negative value which causes `java.lang.ArrayIndexOutOfBoundsException` in `System.arraycopy` method.

I wrote a test case for reproduce this error in `BufferedSocketInputStreamTest`

```java
    @Test
    public void testReadToTheEnd() throws Exception {
        BufferedSocketInputStream in = new BufferedSocketInputStream(new ByteArrayInputStream(new byte[]{
            0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16}), 20);
        assertEquals(in.read(), 0);
        assertEquals(in.read(), 1);
        byte[] buf = new byte[6];
        assertEquals(in.read(buf, 0, buf.length), 6);
        assertEquals(in.read(buf, 0, buf.length), 6);
        assertEquals(in.read(buf, 0, buf.length), 3); // only 3 bytes left in stream
        assertEquals(in.read(buf, 0, buf.length), -1); // got the end of stream, so return -1
    }
```

This test case will throw the same `java.lang.ArrayIndexOutOfBoundsException` as I mentioned above, and I have a patch for fixing this issue, could you help to review it? Thank you a lot.

### PR DESCRIPTION
**Pull request by TengHuo**
add a fix and a test case for ArrayIndexOutOfBoundsException in BufferedSocketInputStream.java:65

### CODE REVIEW

File: src/main/java/com/github/shyiko/mysql/binlog/io/BufferedSocketInputStream.java [hunk 1]
```diff
@@ -61,10 +61,10 @@ public int read(byte[] b, int off, int len) throws IOException {
             offset = 0;
             limit = in.read(buffer, 0, buffer.length);
         }
-        int bytesRemainingInBuffer = Math.min(len, limit - offset);
+        int bytesRemainingInBuffer = Math.min(len, Math.max(limit - offset, 0));
```
Comment (gunnarmorling) [inline src/main/java/com/github/shyiko/mysql/binlog/io/BufferedSocketInputStream.java:64]: How about to return early if `bytesRemainingInBuffer` is 0? No need to copy then or update the offset, right?

**Review by gunnarmorling — COMMENTED** [review]
One suggestion inline.

@osheroff, do you happen to know why whe have that buffered stream to begin with? Couldn't the JDK's own `BufferedInputStream` be used instead?

**Comment by TengHuo** [comment]
> Any updates about this issue?

Sorry for no updating. We have fixed it in Debezium mysql connector internally. Let me update this PR.

**Comment by TengHuo** [comment]
This issue has been fixed by this commit: https://github.com/osheroff/mysql-binlog-connector-java/commit/324858397ec1cb8328d6035f67d411bb16d17740

No need to merge this PR.

**Pull request closed without merging**